sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/JoinEstimation.scala

```diff
@@ -233,7 +233,8 @@ case class JoinEstimation(join: Join) extends Logging {
     val rightKeyStat = rightStats.attributeStats(rightKey)
     val maxNdv = leftKeyStat.distinctCount.get.max(rightKeyStat.distinctCount.get)
     // Compute cardinality by the basic formula.
-    val card = BigDecimal(leftStats.rowCount.get * rightStats.rowCount.get) / BigDecimal(maxNdv)
+    val card = BigDecimal(leftStats.rowCount.get * rightStats.rowCount.get) /
+      BigDecimal(leftKeyStat.distinctCount.get.min(rightKeyStat.distinctCount.get))
```

Comment:
The original formula is practically very rough, but it gives the exact cardinality if the distribution of each table is uniform and all values in the table with smaller distinctCount match in the other table.
But can you share if there is any theory behind the new formula?